Fix http/tests/security/xss-DENIED-object-element.html when running with site isolation enabled

achristensen07 opened the pull request:
#### d7e833adfe65da38ccc1a50b5a55936bf3b9b354
<pre>
Fix http/tests/security/xss-DENIED-object-element.html when running with site isolation enabled
<a href="https://bugs.webkit.org/show_bug.cgi?id=272711">https://bugs.webkit.org/show_bug.cgi?id=272711</a>
<a href="https://rdar.apple.com/126512216">rdar://126512216</a>

Reviewed by Charlie Wolfe.

HTMLPlugInImageElement::canLoadURL needs to return false if there is no content document in this process,
which means that site isolation is on and the document is in another process and isSameOriginDomain is false.

* LayoutTests/platform/mac-site-isolation/TestExpectations:
* Source/WebCore/html/HTMLPlugInImageElement.cpp:
(WebCore::HTMLPlugInImageElement::canLoadURL const):

Canonical link: <a href="https://commits.webkit.org/277543@main">https://commits.webkit.org/277543@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/e425ed27ba9067e03fe25920a238d1bd971393f4

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/47924 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/27125 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/50785 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/50605 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/43977 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/50230 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/32997 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/24612 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/38987 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/48506 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/24788 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/41438 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/20285 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/22272 "Passed tests") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/10/builds/42620 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 wpe-skia](https://ews-build.webkit.org/#/builders/52/builds/5971 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/44269 "Passed tests") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/36/builds/43058 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/52499 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/22966 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/19316 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/46294 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/24236 "Built successfully") | | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/45338 "Passed tests") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/25028 "Built successfully") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/6793 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/23957 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

charliewolfe commented on Source/WebCore/html/HTMLPlugInImageElement.cpp:
```diff
@@ -96,12 +96,15 @@ bool HTMLPlugInImageElement::canLoadURL(const String& relativeURL) const
         RefPtr<Document> contentDocument = this->contentDocument();
-        if (contentDocument && !document().securityOrigin().isSameOriginDomain(contentDocument->securityOrigin()))
+        if (!contentDocument) {
+            ASSERT(document().settings().siteIsolationEnabled());
```
`LocalFrame` holds a RefPtr to document (so it can be null). So I'm not sure this assert will always be correct.

cdumez replied:
What part of the assertion are you worried about?

cdumez replied:
I am not sure I understand why it matters for this assertion that LocalFrame has a RefPtr to the document.

charliewolfe replied:
The assertion assumes we return null in `contentDocument()` because `m_contentFrame` is a RemoteFrame (i.e. site isolation is enabled). But it seems like `localFrame->document()` could be null too.

achristensen07 replied:
This change was incorrect.  There were actually several tests that verified that we can have a null contentDocument without site isolation enabled, and the original version of the PR produced incorrect results.  Updated the PR to check for a RemoteFrame instead.

achristensen07 committed:
Fix http/tests/security/xss-DENIED-object-element.html when running with site isolation enabled
https://bugs.webkit.org/show_bug.cgi?id=272711
rdar://126512216

Reviewed by Charlie Wolfe.

HTMLPlugInImageElement::canLoadURL needs to return false if there is no content document in this process,
which means that site isolation is on and the document is in another process and isSameOriginDomain is false.

* LayoutTests/platform/mac-site-isolation/TestExpectations:
* Source/WebCore/html/HTMLPlugInImageElement.cpp:
(WebCore::HTMLPlugInImageElement::canLoadURL const):

Canonical link: https://commits.webkit.org/277543@main
